### ISSUE TITLE
Bug with reversed behavior on Timer macro action - v 5.2

### ISSUE THREAD

**Issue by midweste** [reef-pi/reef-pi#1811]
After some extensive testing, I've come to find out that in order to get a macro tor run forward as programmed, the timer **Action** needs to be set to **"Turn Off"**.

running 5.2, on Raspberry Pi 3 Model B Plus Rev 1.3

![image](https://user-images.githubusercontent.com/6331501/164319891-121eca4d-87e9-499c-8730-97d28319a56a.png)

![image](https://user-images.githubusercontent.com/6331501/164320038-a354792a-feab-4661-9b41-a4de28870a03.png)

Apr 20 15:40:00 reefpi reef-pi[342]: 2022/04/20 15:40:00 macro-subsystem. Running: Auto Feed
Apr 20 15:40:00 reefpi reef-pi[342]: 2022/04/20 15:40:00 macro-subsystem: executing step:  equipment id: 7  state: false
Apr 20 15:40:00 reefpi reef-pi[342]: 2022/04/20 15:40:00 Equipment: Skimmer On: false
Apr 20 15:40:00 reefpi reef-pi[342]: 2022/04/20 15:40:00 macro-subsystem: executing step: sleep for 60 seconds
Apr 20 15:41:00 reefpi reef-pi[342]: 2022/04/20 15:41:00 macro-subsystem: executing step:  equipment id: 6  state: false
Apr 20 15:41:00 reefpi reef-pi[342]: 2022/04/20 15:41:00 Equipment: Gyre On: false
Apr 20 15:41:00 reefpi reef-pi[342]: 2022/04/20 15:41:00 macro-subsystem: executing step:  equipment id: 8  state: false
Apr 20 15:41:00 reefpi reef-pi[342]: 2022/04/20 15:41:00 Equipment: Pump On: false
Apr 20 15:41:00 reefpi reef-pi[342]: 2022/04/20 15:41:00 macro-subsystem: executing step: sleep for 60 seconds
Apr 20 15:42:00 reefpi reef-pi[342]: 2022/04/20 15:42:00 macro-subsystem: executing step:  macro id: 11  state: true
Apr 20 15:42:00 reefpi reef-pi[342]: 2022/04/20 15:42:00 macro-subsystem. Running: Feeder Trigger
Apr 20 15:42:00 reefpi reef-pi[342]: 2022/04/20 15:42:00 macro-subsystem: executing step:  equipment id: 10  state: true
Apr 20 15:42:00 reefpi reef-pi[342]: 2022/04/20 15:42:00 Equipment: Feeder On: true
Apr 20 15:42:00 reefpi reef-pi[342]: 2022/04/20 15:42:00 macro-subsystem: executing step: sleep for 360 seconds
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 macro-subsystem: executing step:  equipment id: 10  state: false
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 Equipment: Feeder On: false
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 macro-subsystem. Finished: Feeder Trigger
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 macro-subsystem: executing step:  equipment id: 8  state: true
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 Equipment: Pump On: true
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 macro-subsystem: executing step:  equipment id: 6  state: true
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 Equipment: Gyre On: true
Apr 20 15:48:00 reefpi reef-pi[342]: 2022/04/20 15:48:00 macro-subsystem: executing step: sleep for 60 seconds
Apr 20 15:49:00 reefpi reef-pi[342]: 2022/04/20 15:49:00 macro-subsystem: executing step:  equipment id: 7  state: true
Apr 20 15:49:00 reefpi reef-pi[342]: 2022/04/20 15:49:00 Equipment: Skimmer On: true
Apr 20 15:49:00 reefpi reef-pi[342]: 2022/04/20 15:49:00 macro-subsystem. Finished: Auto Feed

**Comment by midweste** [comment]
Likewise, In order to run nested macros in a timer, I needed to set the macros to **"Turn Off"**, and the macros would run in the order I expected them to.

![image](https://user-images.githubusercontent.com/6331501/164324349-70f808d4-8927-451f-b9f1-b5624fcbf8d8.png)

![image](https://user-images.githubusercontent.com/6331501/164324573-615ca198-7d72-4836-83e9-70c1b271aeec.png)

**Comment by midweste** [comment]
Bumping this bug, looks like it fell off the radar

**Comment by ranjib** [comment]
sorry for getting back on this late. I'll review this before the next release.

**Comment by midweste** [comment]
I agree, run forward, run reverse would likely help understanding, however, it does seem they are reversed and Turn Off should mapped to "Run Forward".  In my case above, turn off stay off ran the macro in forward direction (direction specified in the macro). Thanks for responding!

**Comment by borysohnsorge** [comment]
Hi @ranjib,
 similar problem in v6.0

I've made reversible macro to control my refugium light, based on equipment switch. From macro tab everything works as expected. I can succesfully turn the light on (run), and with revert button turn it off. 
![Screenshot 2023-10-18 at 09 19 16](https://github.com/reef-pi/reef-pi.github.io/assets/35881318/ff26484b-003b-41f9-8146-08e2171088e2)

But when I try to use this macro with the timer
![Screenshot 2023-10-18 at 10 08 32](https://github.com/reef-pi/reef-pi.github.io/assets/35881318/05fb16a8-876f-4215-a026-9450c8b4ff60)
my refugium lamp turn on, but it doesn't turn off. 

Here is log from reef-pi:
```
Oct 18 09:28:38 home reef-pi[408719]: 2023/10/18 09:28:38 Successfully added cron entry. ID: 8
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 timer subsystem. Executing module  macro element 6
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem. Running: Refugium
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem: executing step:  equipment id: 16  state: false
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 Equipment: Refugium On: false
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem. Finished: Refugium
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem. Running: Refugium
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem: executing step:  equipment id: 16  state: true
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 Equipment: Refugium On: true
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 macro-subsystem. Finished: Refugium
Oct 18 09:29:05 home reef-pi[408719]: 2023/10/18 09:29:05 timer subsystem. Executing module  macro element 6 reversed
Oct 18 09:30:05 home reef-pi[408719]: 2023/10/18 09:30:05 macro-subsystem. Running: Refugium
Oct 18 09:30:05 home reef-pi[408719]: 2023/10/18 09:30:05 macro-subsystem: executing step:  equipment id: 16  state: true
Oct 18 09:30:05 home reef-pi[408719]: 2023/10/18 09:30:05 Equipment: Refugium On: true
Oct 18 09:30:05 home reef-pi[408719]: 2023/10/18 09:30:05 macro-subsystem. Finished: Refugium
Oct 18 09:30:05 home reef-pi[408719]: 2023/10/18 09:30:05 timer subsystem. Executing module  macro element 6 reversed
Oct 18 09:31:00 home reef-pi[408719]: 2023/10/18 09:31:00 macro-subsystem. Running: Refugium
Oct 18 09:31:00 home reef-pi[408719]: 2023/10/18 09:31:00 macro-subsystem: executing step:  equipment id: 16  state: true
Oct 18 09:31:00 home reef-pi[408719]: 2023/10/18 09:31:00 Equipment: Refugium On: true
Oct 18 09:31:00 home reef-pi[408719]: 2023/10/18 09:31:00 macro-subsystem. Finished: Refugium
Oct 18 09:31:00 home reef-pi[408719]: 2023/10/18 09:31:00 timer subsystem. Executing module  macro element 6 reversed
```

Is there a way to use a reversible macro with a timer, or am I doing something wrong?

When you mention:
> When Marco is "on", all macro steps are executed as defined, in the same order and with the same action.

You mean "Run" - it works fine in the macros tab.

> When the macro is "off", all steps are performed in reverse order (bottom first) and in reverse operation.

You mean "Revert" - this works fine in the macros tab.

> So when the timer triggers the "turn off" action in your macro, it does the opposite of what you defined.

If I understand correctly (last one) and as You can see in my macro, I only have "Turn On" step there. The "turn off" action in the timer should work the opposite of what I defined in this macro. So it should do the same as "Revert" action ("turn off" equipment switch). Unfortunately this does not happen and the lamp remains on.

Regards 
Borys
